app/src/main/java/com/github/se/eventradar/viewmodel/qrCode/ScanTicketQrViewModel.kt

```diff
@@ -0,0 +1,181 @@
+      return hiltViewModel<ScanTicketQrViewModel, Factory>(
+          creationCallback = { factory -> factory.create(eventId = eventId) })
+    }
+  }
+
+  private val _uiState = MutableStateFlow(QrCodeScanTicketState())
+  val uiState: StateFlow<QrCodeScanTicketState> = _uiState
+
+  enum class Action {
+    ScanTicket,
+    ApproveEntry,
+    DenyEntry,
+    FirebaseUpdateError,
+    FirebaseFetchError,
+    AnalyserError,
+  }
+
+  enum class Tab {
+    MyEvent,
+    ScanQr
+  }
+
+  init {
+    qrCodeAnalyser.onDecoded = { decodedString ->
+      val result = decodedString ?: "Failed to decode QR Code"
+      updateDecodedString(result) // Update state flow
+      if (result != "Failed to decode QR Code") {
+        //        println("correctly decoded")
+        updatePermissions(result) // Directly call updateFriendList
+      } else {
+        //      println("wrongly decoded")
+        changeAction(Action.AnalyserError)
+        //        println("checkpoint 2")
+      }
+    }
+    getEventData()
+  }
+
+  private fun updatePermissions(decodedString: String) {
+
+    println("entered updatePermissions")
+    val uiLength = 28
+    val attendeeID = decodedString.take(uiLength)
+    Log.d("QrCodeTicketViewModel", "Ticket User ID: $attendeeID")
+
+    viewModelScope.launch {
+      val attendeeUserDeferred = async { userRepository.getUser(attendeeID) }
+      val currentEventDeferred = async { eventRepository.getEvent(myEventID) }
+
+      val attendeeUser = attendeeUserDeferred.await()
+      val currentEvent = currentEventDeferred.await()
+
+      if (attendeeUser is Resource.Success && currentEvent is Resource.Success) {
+        retryUpdate(attendeeUser.data!!, currentEvent.data!!)
+      } else {
+        changeAction(Action.FirebaseFetchError)
+        return@launch
+      }
+    }
+  }
+
+  private suspend fun retryUpdate(user: User, event: Event) {
+    println("entered retryUpdate")
+    var maxNumberOfRetries = 3
+    var updateResult: Resource<Any>?
+    if (user.eventsAttendeeList.contains(myEventID) && event.attendeeList.contains(user.userId)) {
+      println("both contain one another")
+      user.eventsAttendeeList.remove(myEventID)
+      val userUpdateResult = userRepository.updateUser(user)
+      event.attendeeList.remove(user.userId)
+      val eventUpdateResult = eventRepository.updateEvent(event)
+
+      do {
+        // Check if both updates were successful
+        updateResult =
+            if (userUpdateResult is Resource.Success && eventUpdateResult is Resource.Success) {
+              changeAction(Action.ApproveEntry)
+              Resource.Success(Unit)
+            } else {
+              changeAction(Action.FirebaseUpdateError)
+              Resource.Failure(Exception("Failed to update user and event"))
+            }
+      } while ((updateResult !is Resource.Success) && (maxNumberOfRetries-- > 0))
+    } else {
+      println("one does not contain the other")
```

Comment:
nit: `println` will print to system but that is not very useful. It would be better to use `Log`